broadcast/nosync.go

```diff
@@ -0,0 +1,109 @@
+package broadcast
+
+import (
+	"context"
+)
+
+// NoSyncBroadcaster is a broadcast service that allows a single sender to send
+// messages to multiple recievers. It does not implement any synchronisation
+// and can only be used if the sender is externally synchronised
+// e.g. only used in one goroutine or using a mutex.
+type NoSyncBroadcaster[T any] struct {
+	subscribers map[<-chan T]chan T
+	bufferSize  int
+}
+
+// NewNoSyncBroadcaster creates a NoSyncBroadcaster where all subscribers will
+// get a channel of capacity bufferSize when they subscribe.
+func NewNoSyncBroadcaster[T any](bufferSize int) *NoSyncBroadcaster[T] {
+	return &NoSyncBroadcaster[T]{
+		subscribers: make(map[<-chan T]chan T),
+		bufferSize:  bufferSize,
+	}
+}
+
+// Subscribe creates and returns a new channel that will receive all messages
+// send by the sender via this broadcast service.
+func (b *NoSyncBroadcaster[T]) Subscribe() <-chan T {
+	ch := make(chan T, b.bufferSize)
+	b.AddSubscriber(ch)
+	return ch
+}
+
+// AddSubscriber gives subscribers the option to provide their own chanel to
+// receive updates on. In case they already have allocated one and want to
+// reuse it or if the default bufferSize isn't OK for them.
+func (b *NoSyncBroadcaster[T]) AddSubscriber(sub chan T) {
+	b.subscribers[sub] = sub
+}
+
+// Unsubscribe will stop the service sending messages on this channel and close
+// the channel. Returns true if the provided channel is a valid subscriber.
+func (b *NoSyncBroadcaster[T]) Unsubscribe(sub <-chan T) bool {
+	if sub, ok := b.subscribers[sub]; ok {
+		close(sub)
+		delete(b.subscribers, sub)
+		return true
+	}
+	return false
+}
+
+// CloseAll will close and delete all subscribers' channels.
+func (b *NoSyncBroadcaster[T]) CloseAll() {
+	for _, sub := range b.subscribers {
```

Comment:
No, it's not OK to use `map[chan T]struct{}` instead of `map[<-chan T]chan T`. We're giving out read-only channels and those cannot be transformed back into full channels and therefore couldn't be used to index the map. Full channels, however can be implicitly converted into read-only channels and can be used to delete by-key in the second scenario.

```go
m := make(map[chan int]struct{})
c := make(chan int)
m[c] = struct{}{}
var ro <-chan int
ro = c
m[ro] // compile error cannot use <-chan int to index map[chan int]struct{}

for sub := range m {
    if sub == ro {
        fmt.Println("Found it!")
    }
}
```

So using `map[<-chan T]chan T` is a quick way of getting a read-only to full channel conversion, that can be used to close the channel when the user unsubscribes.

Reply:
Another important note about the above is that read-only channels cannot be closed. Only a holder of the full channel can do that.